### PR TITLE
Add functions for creating and changing networks

### PR DESCRIPTION
**Pull request by AbelHeinsbroek**
Added ENaddnode, ENaddlink, ENremovenode and ENremovelink functions to change the network.
Added ENinit function to start a new network from scratch

### CODE REVIEW

File: src/epanet.c [hunk 1]
```diff
@@ -3166,6 +3302,352 @@ int DLLEXPORT ENgetaveragepatternvalue(int index, EN_API_FLOAT_TYPE *value)
...
+  int i, linkType;
+  
+  if (!Openflag) return(102);
+  if (index <= 0 || index > Nlinks) return(203);
+  
+  ENgetlinktype(index, &linkType);
+  
+  // remove from hash table
+  ENHashTableDelete(LinkHashTable, Link[index].ID);
+  ENHashTableFind(LinkHashTable, Link[index].ID);
```
Comment (samhatchett) [inline src/epanet.c]: probably want to check for error on delete, then ensure that `find` returns `NOTFOUND` ?

File: src/epanet.c [hunk 2]
```diff
@@ -3166,6 +3302,352 @@ int DLLEXPORT ENgetaveragepatternvalue(int index, EN_API_FLOAT_TYPE *value)
...
+  return(0);
+}
+
+int DLLEXPORT ENaddlink(char *id, EN_LinkType linkType, char *fromNode, char *toNode)
+{
+  int i, n;
+  
+/* Check if a link with same id already exists */
+  if ( !Openflag ) return(102);
+  if ( ENgetlinkindex(id, &i) == 0 ) return(215);
```
Comment (samhatchett) [inline src/epanet.c]: see summary note on conditional coding style. non-breaking conditional on an error code makes it hard to debug.

File: src/epanet.c [hunk 3]
```diff
@@ -3166,6 +3302,352 @@ int DLLEXPORT ENgetaveragepatternvalue(int index, EN_API_FLOAT_TYPE *value)
...
+  Nlinks++;
+  n = Nlinks;
+  
+/* Grow arrays to accomodate the new value */
+  Link = (Slink *)realloc(Link, (Nlinks+1)*sizeof(Slink));
+  Q = (double *)realloc(Q, (Nlinks+1)*sizeof(double));
+  LinkSetting = (double *)realloc(LinkSetting, (Nlinks+1)*sizeof(double));
+  LinkStatus = (char *)realloc(LinkStatus, (Nlinks+1)*sizeof(char));
+  
+  strncpy(Link[n].ID, id, MAXID);
```
Comment (samhatchett) [inline src/epanet.c]: Edge case here: if you specify a link with an ID of more than `MAXID` (31) characters, then the name will be truncated and you won't be able to access the new link by the ID you specified. How to handle?

File: src/hash.c [hunk 4]
```diff
@@ -72,13 +73,54 @@ int ENHashTableInsert(ENHashTable *ht, char *key, int data)
   return(1);
 }
 
+/* Abel Heinsbroek: Added function to update the hash table value for a given key */
```
Comment (samhatchett) [inline src/hash.c]: attribution in the code is not needed. Git log takes care of this.

File: src/funcs.h [hunk 5]
```diff
@@ -43,6 +43,9 @@ double  interp(int, double *,             /* Interpolates a data curve  */
                double *, double);
 int     findnode(char *);                 /* Finds node's index from ID */
 int     findlink(char *);                 /* Finds link's index from ID */
+int     findtank(int);                    /* Find tank index from node index */  // (AH)
+int     findvalve(int);                   /* Find valve index from node index */ // (AH)
+int     findpump(int);                    /* Find pump index from node index */  // (AH)
```
Comment (samhatchett) [inline src/funcs.h]: attribution in the code is not needed. Git log takes care of this.

File: src/epanet.c [hunk 6]
```diff
@@ -3166,6 +3302,352 @@ int DLLEXPORT ENgetaveragepatternvalue(int index, EN_API_FLOAT_TYPE *value)
...
+  if(fromType <= EN_PIPE) {
+    Npipes--;
+  } else if(fromType == EN_PUMP) {
+    Npumps--;
+    
+  }
+  
+}
+
+int  DLLEXPORT  ENaddnode(char *id, EN_NodeType nodeType)
```
Comment (samhatchett) [inline src/epanet.c]: see note for links also. Edge case is specifying an ID of length greater than `MAXID` (31) characters, which would truncate the ID silently. Should probably return an error if the user attempts this.

**Comment by samhatchett** [comment]
@AbelHeinsbroek and others - I'd like to suggest just merging this into OWA:dev-net-builder so that we can create issues around desired changes and others can lend a hand.

**Comment by eladsal** [comment]
@samhatchett yes, that would be a good start.

**Comment by samhatchett** [comment]
Thanks again to @AbelHeinsbroek for the contribution! Let's file any desired extensions or revisions into Issues. I'm going to self-assign some easier ones to get the ball rolling.

**Pull request merged**